File: js/setCoveringProblemGreedy.js
function setCoveringProblemGreedy(stations, areas){
	const unAreas = new Set(areas);
	const selectedAreas = [];

	while(unAreas.size != 0){
		let bestStation;
		let coveredAreas = new Set();
		for(let station of stations){
			let covered = new Set([...unAreas].filter(area => station.area.includes(area)));
			if(covered.size > coveredAreas.size){
				bestStation = station;
				coveredAreas = covered;
			}
		}

		if(bestStation){
			selectedAreas.push(bestStation);
			coveredAreas.forEach(area => unAreas.delete(area));
		}
	}

	return selectedAreas;
}

const stations = [
	{name: "Station A", area: ["area1","area2","area3"]},
	{name: "Station B", area: ["area4","area5","area7"]},
	{name: "Station C", area: ["area1","area2","area4"]},
	{name: "Station D", area: ["area4", "area5", "area6"]}
];

const areas = ["area1", "area2", "area3", "area4", "area5", "area6", "area7"]

const result = setCoveringProblemGreedy(stations, areas);

console.log(result);

File: paython/setCoveringProblemGreedy.py
from pprint import pprint

def set_covering_problem_greedy(stations, areas):
    unAreas = set(areas)
    selected_areas = []
    while unAreas:
        best_station = None
        covered_areas = set()
        for station in stations:
            covered = unAreas & set(station["area"])
            if(len(covered) > len(covered_areas)):
                best_station = station
                covered_areas = covered
        if best_station:
            selected_areas.append(best_station)
            unAreas -= covered_areas
    return selected_areas

stations = [
        {"name": "Station A", "area": ["area1","area2","area3"]},
        {"name": "Station B", "area": ["area4","area5", "area6"]},
        {"name": "Station C", "area": ["area1","area6", "area3"]},
        {"name": "Station D", "area": ["area2","area4", "area5", "area7"]}
]

areas = ["area1", "area2", "area3", "area4", "area5", "area6","area7"]

result = set_covering_problem_greedy(stations, areas)
pprint(result)
